src/core/ucc_team.c

```diff
@@ -229,3 +297,105 @@ ucc_status_t ucc_team_destroy(ucc_team_h team)
+    ucc_assert(id >= 1);
+    local[map_pos] |= ((uint64_t)1 << pos);
+}
+
+static ucc_status_t ucc_team_alloc_id(ucc_team_t *team)
+{
+    /* at least 1 ctx is always available */
+    ucc_context_t   *ctx      = team->contexts[0];
+    ucc_tl_iface_t  *tl_iface = UCC_TL_TEAM_IFACE(team->service_team);
+    uint64_t        *local, *global;
+    ucc_status_t     status;
+    int              pos, i;
+
+    if (team->id > 0) {
+        ucc_assert(UCC_TEAM_ID_IS_EXTERNAL(team));
+        return UCC_OK;
+    }
+    ucc_assert(team->service_team);
+    if (!ctx->ids.pool) {
+        ctx->ids.pool = ucc_malloc(ctx->ids.pool_size*2*sizeof(uint64_t), "ids_pool");
+        if (!ctx->ids.pool) {
+            ucc_error("failed to allocate %zd bytes for team_ids_pool",
+                      ctx->ids.pool_size*2*sizeof(uint64_t));
+            return UCC_ERR_NO_MEMORY;
+        }
+        /* init all bits to 1 - all available */
+        memset(ctx->ids.pool, 255, ctx->ids.pool_size*2*sizeof(uint64_t));
+    }
+    local  = ctx->ids.pool;
+    global = ctx->ids.pool + ctx->ids.pool_size;
+
+    if (!team->task) {
+        ucc_tl_team_subset_t subset = {
+            .map.type   = UCC_EP_MAP_FULL,
+            .map.ep_num = team->params.oob.participants,
+            .myrank     = team->rank
+        };
+        status = tl_iface->scoll.allreduce(
+            &team->service_team->super, local, global, UCC_DT_UINT64, ctx->ids.pool_size, UCC_OP_BAND,
+            subset, &team->task);
+        if (status < 0) {
+            ucc_error("failed to start service allreduce for team ids pool allocation: %s",
+                      ucc_status_string(status));
+            return status;
+        }
+    }
+    ucc_context_progress(ctx);
+    status = tl_iface->scoll.test(team->task);
+    if (status < 0) {
+        ucc_error("service allreduce test failure: %s",
+                  ucc_status_string(status));
+        return status;
+    } else if (status != UCC_OK) {
+        return status;
+    }
+    tl_iface->scoll.cleanup(team->task);
+    team->task = NULL;
+    memcpy(local, global, ctx->ids.pool_size*sizeof(uint64_t));
+    pos = 0;
+    for (i=0; i<ctx->ids.pool_size; i++) {
+        if ((pos = find_first_set_and_zero(&local[i])) > 0) {
+            break;
+        }
+    }
+    if (pos > 0) {
+        ucc_assert(pos <= 64);
+        team->id = (uint16_t)(i*64+pos);
+        ucc_info("allocated ID %d for team %p", team->id, team);
+    } else {
+        ucc_warn("could not allocate team id, whole id space is occupied, "
+                 "try increasing UCC_TEAM_IDS_POOL_SIZE");
+        return UCC_ERR_NO_RESOURCE;
+    }
+    ucc_assert(team->id > 0);
+    return UCC_OK;
+}
+
+static void ucc_team_relase_id(ucc_team_t *team)
+{
+    ucc_context_t *ctx = team->contexts[0];
+    /* release the id pool bit if it was not provided by user */
+    if (!UCC_TEAM_ID_IS_EXTERNAL(team)) {
+        set_id_bit(ctx->ids.pool, team->id);
+    }
+}
```

Comment:
Val, I have been thinking about this algorithm. From what I see, this algorithm can cause a live-lock in a multithreaded case. In UCC’s case, we avoid this because we don’t support multiple team posts. Is my current statement correct? If so, should we warn somewhere. Is there an edge case, where we can still cause livelock?